### PR TITLE
Adds cleanup of orphaned OpenFGA tuples.

### PR DESCRIPTION
**Pull request by alesstimec**
## Description

The what and why - include a summary of the change, describe what it does, and include relevant motivation and context.

Fixes _JIRA/GitHub issue number_

## Engineering checklist
*Check only items that apply*

- [ ] Documentation updated
- [ ] Covered by unit tests
- [ ] Covered by integration tests

### CODE REVIEW

File: internal/jimm/access.go
Comment (SimoneDutto) [inline internal/jimm/access.go]: if you are getting the last tuples and you have no orphaned tuples you continue and requery the tuples from the start. Right?

in a case of cleaned db you are potentially spinning and requery continuously and not once every 6hours

File: internal/jimm/access_test.go
Comment (SimoneDutto) [inline internal/jimm/access_test.go:1203]: can you add another `OpenFGACleanup` on a cleaned db, to make sure we are checking the scenario I've described above?

File: internal/jimm/access.go
Comment (kian99) [inline internal/jimm/access.go]: Shouldn't this be `if continuationToken != ""`?

File: internal/jimm/access_test.go
Comment (kian99) [inline internal/jimm/access_test.go:1140]: Something worthwhile for creating OpenFGA tuples would be to create a similar construct to our database `yaml` parser where we define a yaml string of what will go into our database. We could have the same for OpenFGA tuples, define it in a yaml string above the test and call some test function to create them.

**Review by kian99 — DISMISSED** [review]
2 comments.
Overall I think this is fine, but my main gripe is that it's a bit of a bandaid over a different problem. If we're not cleaning up tuples, we're doing something wrong elsewhere and then creating a need for this routine.

Just a thought but perhaps we should log when we have orphaned tuples, then we can see we're not cleaning up when we should. We can already see this when we run `jimmctl auth relation list`.

File: internal/jimm/access.go
Comment (alesstimec) [reply internal/jimm/access.go]: fixed, thanks

File: internal/jimm/access.go
Comment (alesstimec) [reply internal/jimm/access.go]: yes, fixed

File: internal/jimm/access_test.go
Comment (alesstimec) [reply internal/jimm/access_test.go:1140]: i could do, but i wanted to create an arbitrary number of tuples so a function seemed to make more sense

**Comment by alesstimec** [comment]
I agree this is a band-aid for a problem we have in JIMM. I've added logging for tuples that are removed so that we can see where the problem is and address it in followups.

File: internal/jimm/access_test.go
Comment (alesstimec) [reply internal/jimm/access_test.go:1203]: added

**Review by SimoneDutto — APPROVED** [review]
lgtm

File: internal/jimm/access.go
Comment (ale8k) [inline internal/jimm/access.go:848]: I'd godoc this to explain we're using ToJAASTag specifically to check for existence and it just so happens this method works for us. The code isn't clear without knowing this.

File: internal/jimm/access.go
Comment (ale8k) [inline internal/jimm/access.go]: 10 feels like a bit of a magic number, what is the reason for doing it in 10s?

**Review by ale8k — COMMENTED** [review]
I want a little time to think about the "removing if we can't find it in db". 

For example, if we had data loss [somehow; just humour me] in the db. And we just so happened to be performing a cleanup, we'll remove access once data is restored.

And I'll bet there's other scenarios we're not thinking of. So before merging this I implore we all have a think about it.

I'm also unsure about the reading in 10s, and want to be make sure it isn't rate limited. 

Again humour me, we have a data loss of 10000 objects, OpenFGA begins to reject requests after the first 5000, data is restored, we have 5000 lingering, data is restored and 5000 tuples are valid leaving our auth in a mess. (Not sure if this could actually happen but I'm throwing ideas to the wind).

I guess what I'm trying to say is let's have a think and be careful before merging.

File: internal/jimm/access.go
Comment (alesstimec) [reply internal/jimm/access.go]: i can change it to 17?

File: internal/jimm/access.go
Comment (kian99) [inline internal/jimm/access.go]: This log line seems arbitrary and at the wrong log level.
I'd suggest to either remove it or add some info
`zapctx.Debug(ctx, "continue cleanup",zap.Int("page",page)`

File: internal/jimm/access.go
Comment (kian99) [inline internal/jimm/access.go:813]: Let's update the godoc to make it very explicit that this cleanup method is intended to be short-lived until tuples are correctly removed from OpenFGA. Something like.
```
//....
// 
// This approach to cleaning up tuples is intended to be temporary while we implement
// a better approach to eventual consistency of JIMM's database objects and OpenFGA tuples.
```

File: internal/jimm/access.go
Comment (kian99) [inline internal/jimm/access.go]: We're not passing the continuation token into the function?

File: internal/jimm/access.go
Comment (kian99) [inline internal/jimm/access.go]: Seems to be the wrong way to check for termination, just check if the continuation token is empty and if so, return. Which is already being done below.

File: internal/jimm/access_test.go
Comment (kian99) [inline internal/jimm/access_test.go]: I think we need a test with more than 10 tags to make sure pagination is working correctly.

File: internal/jimm/access.go
Comment (alesstimec) [reply internal/jimm/access.go]: removed.

File: internal/jimm/access.go
Comment (alesstimec) [reply internal/jimm/access.go]: added

File: internal/jimm/access.go
Comment (alesstimec) [reply internal/jimm/access.go]: changed/fixed

File: internal/jimm/access_test.go
Comment (alesstimec) [reply internal/jimm/access_test.go]: done

File: internal/jimm/access.go
Comment (alesstimec) [reply internal/jimm/access.go:813]: done

**Comment by SimoneDutto** [comment]
fyi: i was curious about @ale8k comments and I've checked.

In case of a data loss Postgres enters a recovery mode (read-only), using the WAL to replay the latest operations.
So in case of a data loss, I haven't found any documentation that indicates that Postgres can answer your queries w/o errors but with records missing.

File: internal/jimm/access.go
Comment (kian99) [inline internal/jimm/access.go]: This log line is logging **all** tuples, not the ones that are orphaned.

**Review by kian99 — COMMENTED** [review]
Looks good, will approve after the log comment is addressed.

File: internal/jimm/access.go
Comment (alesstimec) [reply internal/jimm/access.go]: done

**Review by kian99 — APPROVED** [review]
A suggestion for a follow-up. Add a metric to measure the time on the orphan relation cleanup method.

**Pull request merged**
